File: binary_search/binary_search_iterations.rb
# 0(log n) time complexity
def binary_search_iterations(list, item)
  low = 0 #first index in array
  high = list.length - 1 #last index of array
  iterations = 0 #will increment each time search is run unsuccessfully

  while low <= high
    mid = (low + high)/2 #check middle element
    guess = list[mid] 

    if guess == item
      return "it took #{iterations} iterations to guess the number"
    elsif guess > item
      high = mid-1
      iterations += 1
      p "Lower than #{mid}"
    else
      low = mid + 1
      iterations += 1
      p "Higher than #{mid}"
    end
  end
end

my_list = (1..100).to_a
puts binary_search_iterations(my_list, 20)

File: binary_search/binary_search_iterations.js
//Count number of iterations it takes to find number using binary search
//0(log n)

function binarySearchIterations(list, item) {
  let low = 0;
  let high = list.length - 1;
  let mid;
  let iterations = 0;

  while(low <= high) {
    mid = Math.floor((low + high) / 2);
    let guess = list[mid];

    if(guess === item) {
      return `It took ${iterations} iterations`
    } else if (guess > item) {
      high = mid - 1;
      iterations += 1;
      console.log(`Lower than ${mid}`)
    } else {
      low = mid + 1;
      iterations += 1;
      console.log(`Higher than ${mid}`)
    }
  }
  return null;
};

myList = Array.from(Array(100).keys());
console.log(binarySearchIterations(myList, 20))
console.log(binarySearchIterations(myList, 67))

